pkg/asset/asset.go

```diff
@@ -1,13 +1,56 @@
 package asset
 
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
 // Asset used to install OpenShift.
 type Asset interface {
 	// Dependencies returns the assets upon which this asset directly depends.
 	Dependencies() []Asset
 
-	// Generate generates this asset given the states of its dependent assets.
-	Generate(map[Asset]*State) (*State, error)
+	// Generate generates this asset given the states of its parent assets.
+	Generate(Parents) error
 
 	// Name returns the human-friendly name of the asset.
 	Name() string
 }
+
+// WritableAsset is an Asset that has files that can be written to disk.
+type WritableAsset interface {
+	Asset
+
+	// Files returns the files to write.
+	Files() []*File
+}
+
+// File is a file for an Asset.
+type File struct {
+	// Filename is the name of the file.
+	Filename string
+	// Data is the contents of the file.
+	Data []byte
+}
+
+// PersistToFile writes all of the files of the specified asset into the specified
+// directory. If the asset is not a WritableAsset, then no files will be written.
+func PersistToFile(asset Asset, directory string) error {
+	writableAsset, ok := asset.(WritableAsset)
```

Comment:
Why not just change the signature of `PersistToFile`? Targets shouldn't include non-writable assets.

Reply:
That is fine, too. I'll make that change.